### ISSUE TITLE
Will there be custom GPU kernels for NHWC Batchnorm in the future?

### ISSUE THREAD

**Issue by ghost** [tensorflow/tensorflow#2377]
NCHW Batchnorm #1759 is one merge away but as it currently stands the NHWC Batchnorm kernel is the single biggest performance killer in very deep networks. My testing shows that Resnet-164's forward pass taking more than twice as much time as that of Resnet-110, simply because the former has an additional BN of increased channel dimension (x4) in each block, despite the two architecture having nearly the same FLOPs.

Edit: Actually meant Resnet-164, not Resnet-152

**Comment by zheng-xq** [comment]
NHWC will be supported by converting to NCHW internally back and forth. Our experience is that in most cases that is okay. If you absolutely don't want that overhead, you can switch your entirely model to NCHW through data_format, and that will be fine. If for some ops, NCHW support isn't available, either a conversion can be added, or add new support.

**Comment by girving** [comment]
@zheng-xq: Should I leave this assigned to you, or is it already done?

**Comment by zheng-xq** [comment]
We are still waiting for the BatchNorm kernel to be merged. Support for other formats will proceed afterwards.

**Comment by andydavis1** [comment]
Chatted with @zheng-xq offline: someone on the team is looking at this.

**Comment by thuyen** [comment]
What would be the timeline for the BatchNorm kernel to be merged?

**Comment by drpngx** [comment]
Not sure if it's obsolete. Closing since it might have been done. Feel free to open a new github issue if the problem still persists in recent versions.
